File: generate_embeddings.py
#!/usr/bin/env python3
"""
Generate embeddings for semantic search
Runs during Jekyll build to create search-embeddings.json
Uses sentence-transformers for zero-cost, offline embeddings
"""

import json
import os
from pathlib import Path
from sentence_transformers import SentenceTransformer

def load_search_data():
    """Load existing search data from Jekyll build"""
    search_data_path = Path('_site/assets/js/search-data.json')

    if not search_data_path.exists():
        print("❌ Error: search-data.json not found. Run 'bundle exec jekyll build' first!")
        return None

    with open(search_data_path, 'r', encoding='utf-8') as f:
        return json.load(f)

def generate_embeddings(search_data):
    """Generate embeddings for all pages"""
    print("🚀 Loading embedding model (all-MiniLM-L6-v2)...")
    print("   (First run will download ~90MB model, then cached)")

    # Load lightweight, fast model
    model = SentenceTransformer('all-MiniLM-L6-v2')

    print(f"\n📝 Generating embeddings for {len(search_data)} pages...")

    embeddings_data = []

    for i, page in enumerate(search_data, 1):
        # Combine title and content for better semantic understanding
        text = f"{page['title']} {page.get('content', '')}"

        # Generate embedding (384-dimensional vector)
        embedding = model.encode(text).tolist()

        embeddings_data.append({
            'id': page['id'],
            'title': page['title'],
            'url': page['url'],
            'content': page.get('content', ''),
            'heading': page.get('heading', ''),
            'embedding': embedding
        })

        print(f"   [{i}/{len(search_data)}] {page['title']}")

    return embeddings_data

def save_embeddings(embeddings_data):
    """Save embeddings to JSON file"""
    output_path = Path('_site/assets/js/search-embeddings.json')
    output_path.parent.mkdir(parents=True, exist_ok=True)

    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(embeddings_data, f, ensure_ascii=False)

    # Calculate file size
    size_kb = output_path.stat().st_size / 1024
    print(f"\n✅ Embeddings saved to: {output_path}")
    print(f"   File size: {size_kb:.1f} KB")

def main():
    print("=" * 60)
    print("🔍 Semantic Search Embedding Generator")
    print("=" * 60)

    # Load search data
    search_data = load_search_data()
    if not search_data:
        return

    # Generate embeddings
    embeddings_data = generate_embeddings(search_data)

    # Save to file
    save_embeddings(embeddings_data)

    print("\n" + "=" * 60)
    print("✨ Done! Embeddings are ready for semantic search")
    print("=" * 60)
    print("\nNext steps:")
    print("  1. Embeddings are now in _site/assets/js/search-embeddings.json")
    print("  2. Chatbot will automatically use them for semantic search")
    print("  3. Re-run this script whenever you update content")

if __name__ == '__main__':
    main()

File: _site_vercel/fachandbook/generate-embeddings.js
/**
 * Generate semantic search embeddings using Transformers.js
 * Zero API calls - runs completely offline!
 * Uses the same model as Python sentence-transformers
 */

import { pipeline } from '@xenova/transformers';
import fs from 'fs';
import path from 'path';

console.log('='.repeat(60));
console.log('🔍 Semantic Search Embedding Generator (Node.js)');
console.log('='.repeat(60));

async function loadSearchData() {
  const searchDataPath = path.join('_site', 'assets', 'js', 'search-data.json');

  if (!fs.existsSync(searchDataPath)) {
    console.error('❌ Error: search-data.json not found.');
    console.error('   Run "bundle exec jekyll build" first!');
    return null;
  }

  const data = JSON.parse(fs.readFileSync(searchDataPath, 'utf-8'));
  console.log(`\n✓ Loaded ${Object.keys(data).length} pages from search-data.json`);
  return data;
}

async function generateEmbeddings(searchData) {
  console.log('\n🚀 Loading embedding model (all-MiniLM-L6-v2)...');
  console.log('   (First run will download ~100MB model, then cached)');

  // Load the same model used in the browser
  const extractor = await pipeline('feature-extraction', 'Xenova/all-MiniLM-L6-v2');

  console.log('✓ Model loaded!\n');

  const embeddingsData = [];
  const entries = Object.entries(searchData);
  const total = entries.length;

  console.log(`📝 Generating embeddings for ${total} pages...\n`);

  for (let i = 0; i < entries.length; i++) {
    const [pageId, page] = entries[i];

    // Combine title and content for better semantic understanding
    const text = `${page.title || ''} ${page.content || ''}`;

    // Generate embedding
    const output = await extractor(text, { pooling: 'mean', normalize: true });
    const embedding = Array.from(output.data);

    embeddingsData.push({
      id: pageId,
      title: page.title || '',
      url: page.url || '',
      content: page.content || '',
      heading: page.heading || '',
      embedding: embedding
    });

    // Progress indicator
    const percent = Math.round(((i + 1) / total) * 100);
    const bar = '█'.repeat(Math.floor(percent / 2)) + '░'.repeat(50 - Math.floor(percent / 2));
    process.stdout.write(`\r   [${bar}] ${percent}% (${i + 1}/${total}) ${page.title || 'Untitled'}`);
  }

  console.log('\n');
  return embeddingsData;
}

function saveEmbeddings(embeddingsData) {
  const outputPath = path.join('_site', 'assets', 'js', 'search-embeddings.json');
  const outputDir = path.dirname(outputPath);

  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  fs.writeFileSync(outputPath, JSON.stringify(embeddingsData), 'utf-8');

  const stats = fs.statSync(outputPath);
  const sizeKB = (stats.size / 1024).toFixed(1);

  console.log(`✅ Embeddings saved to: ${outputPath}`);
  console.log(`   File size: ${sizeKB} KB`);
  console.log(`   Total embeddings: ${embeddingsData.length}`);
}

async function main() {
  try {
    // Load search data
    const searchData = await loadSearchData();
    if (!searchData) {
      process.exit(1);
    }

    // Generate embeddings
    const embeddingsData = await generateEmbeddings(searchData);

    // Save to file
    saveEmbeddings(embeddingsData);

    console.log('\n' + '='.repeat(60));
    console.log('✨ Done! Embeddings are ready for semantic search');
    console.log('='.repeat(60));
    console.log('\nNext steps:');
    console.log('  1. Embeddings are in _site/assets/js/search-embeddings.json');
    console.log('  2. Chatbot will use them for hybrid semantic + keyword search');
    console.log('  3. Re-run this script whenever you update content');
    console.log('\n🎉 Zero API calls! Completely offline! Free forever!');

  } catch (error) {
    console.error('\n❌ Error:', error.message);
    process.exit(1);
  }
}

main();
